### ISSUE TITLE
Discriminators don't work with enum values

### ISSUE THREAD

**Issue by jmcnevin** [oapi-codegen/oapi-codegen#773]
In testing 96482c12b, I've found that the `From...` and `Merge...` methods generated by discriminator usage do not use the type generated from usage of `enum` in the spec.

For example, my discriminator is a property named `protocol`, with `type: string` with enum values of "HTTPS, TCP, UDP".  The generator creates a custom type for the protocol field in structs like `LoadBalancerHTTPSProtocol`.

```go
type LoadBalancerHttpsProtocol string

const (
	LoadBalancerHttpsProtocolHTTPS LoadBalancerHttpsProtocol = "HTTPS"
	LoadBalancerHttpsProtocolTCP LoadBalancerHttpsProtocol = "TCP"
	LoadBalancerHttpsProtocolUDP LoadBalancerHttpsProtocol = "UDP"
)
```

Later, in generated `Merge...` and `From...` methods, we can see lines like `v.Protocol = "HTTPS"`, which will not compile, because `"HTTPS"` needs to use the proper type.

**Comment by jmelching** [comment]
Yeah, when using anyOf with a discriminator based on an enum, the generator doesn't produce code that compiles. This leaves one to define the discriminator as a string.

**Comment by john-tipper** [comment]
I think this is related to https://github.com/deepmap/oapi-codegen/issues/1245. I had similar issues and resolved it like this: https://github.com/deepmap/oapi-codegen/issues/1245#issuecomment-2014725258

TLDR: ensure all fields used as discriminators are all marked as required in all cases. Then the generated code does not use a pointer for the struct field, meaning it's possible to instantiate them by directly assigning a string literal.

**Comment by mromaszewicz** [comment]
This is a comment from an automated bug scrub, after making a significant number of fixes to the code generation.

We reproduced this against the current `main` and confirmed the compile error is gone.

**Spec used** (`spec.yaml`):
```yaml
openapi: 3.0.3
info:
  title: test
  version: 1.0.0
paths:
  /lb:
    get:
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/LoadBalancer'
components:
  schemas:
    LoadBalancerHttps:
      type: object
      required: [protocol]
      properties:
        protocol: { type: string, enum: [HTTPS, TCP, UDP] }
        httpsField: { type: string }
    LoadBalancerTcp:
      type: object
      required: [protocol]
      properties:
        protocol: { type: string, enum: [HTTPS, TCP, UDP] }
        tcpField: { type: string }
    LoadBalancer:
      oneOf:
        - $ref: '#/components/schemas/LoadBalancerHttps'
        - $ref: '#/components/schemas/LoadBalancerTcp'
      discriminator:
        propertyName: protocol
        mapping:
          HTTPS: '#/components/schemas/LoadBalancerHttps'
          TCP: '#/components/schemas/LoadBalancerTcp'
```

**Config used** (`config.yaml`):
```yaml
package: repro
output: types.gen.go
generate:
  models: true
```

**Result:** the `From…`/`Merge…` methods no longer assign a bare string to the typed discriminator field (the old `v.Protocol = "HTTPS"` that failed to compile). They now marshal the value and inject the discriminator via `runtime.JSONMerge`:
```go
func (t *LoadBalancer) FromLoadBalancerHttps(v LoadBalancerHttps) error {
	b, err := json.Marshal(v)
	if err != nil {
		return err
	}
	b, err = runtime.JSONMerge(b, []byte(`{"protocol":"HTTPS"}`))
	t.union = b
	return err
}
```
Because `protocol` is a string enum, injecting the JSON string `"HTTPS"` is correct, and the generated code **compiles** (verified with `go build` against the current runtime). Closing as fixed.
